components/ntp_background_images/browser/ntp_background_images_service.cc

```diff
@@ -6,82 +6,417 @@
+    if (brave::BraveReferralsService::IsDefaultReferralCode(code)) {
+      MarkThisInstallIsNotSuperReferralForever();
+    } else {
+      // If current code is not an default one, let's check it after fetching
+      // mapping table.
+      DownloadSuperReferralMappingTable();
+    }
+    return;
+  }
+
+  DVLOG(2) << __func__ << ": This has invalid component info.";
+  DVLOG(2) << __func__ << ": In this case, this install is campaign ended super"
+                       << " referral, default referral or non super referral.";
+}
+
+void NTPBackgroundImagesService::MonitorReferralPromoCodeChange() {
+  DVLOG(2) << __func__ << ": Monitor referral promo code change";
+
+  pref_change_registrar_.Init(local_pref_);
+  pref_change_registrar_.Add(kReferralPromoCode,
+      base::BindRepeating(&NTPBackgroundImagesService::OnPreferenceChanged,
+      base::Unretained(this)));
+}
+
+void NTPBackgroundImagesService::OnPreferenceChanged(
+  const std::string& pref_name) {
+  DCHECK_EQ(kReferralPromoCode, pref_name);
+  const std::string new_referral_code = GetReferralPromoCode();
+  DVLOG(2) << __func__ << ": Got referral promo code: "
+                       << new_referral_code;
+  DCHECK(!new_referral_code.empty());
+  if (brave::BraveReferralsService::IsDefaultReferralCode(new_referral_code)) {
+    DVLOG(2) << __func__ << ": This has default referral promo code.";
+    MarkThisInstallIsNotSuperReferralForever();
+    return;
   }
+
+  DVLOG(2) << __func__ << ": This has non default referral promo code."
+                       << " Let's check this code is super referral or not"
+                       << " after downloading mapping table.";
+  DownloadSuperReferralMappingTable();
 }
 
-NTPBackgroundImagesService::~NTPBackgroundImagesService() {}
+void NTPBackgroundImagesService::RegisterSuperReferralComponent() {
+  DVLOG(2) << __func__ << ": Register NTP SR component";
+
+  std::string public_key;
+  std::string id;
+  std::string theme_name;
+  if (initial_sr_component_info_.is_dict()) {
+    public_key = *initial_sr_component_info_.FindStringKey(kPublicKey);
+    id = *initial_sr_component_info_.FindStringKey(kComponentID);
+    theme_name = *initial_sr_component_info_.FindStringKey(kThemeName);
+  } else {
+    const auto* value = local_pref_->Get(
+        prefs::kNewTabPageCachedSuperReferralComponentInfo);
+    public_key = *value->FindStringKey(kPublicKey);
+    id = *value->FindStringKey(kComponentID);
+    theme_name = *value->FindStringKey(kThemeName);
+  }
+
+  RegisterNTPBackgroundImagesComponent(
+      component_update_service_,
+      public_key, id,
+      base::StringPrintf("NTP Super Referral (%s)",
+                         theme_name.c_str()),
+      base::BindRepeating(&NTPBackgroundImagesService::OnComponentReady,
+                          weak_factory_.GetWeakPtr(),
+                          true));
+}
+
+void NTPBackgroundImagesService::DownloadSuperReferralMappingTable() {
```

Comment:
why are we making a separate download for this? This is exactly the kind of thing we don't want to do. This should go in a known component id so the component updater can handle it

Reply:
I think we don't need to use another component for fetching mapping table.
The reason is browser doesn't need to get mapping table update during the browser lifetime. Mapping table is only used at the first run to know whether this install is SR or not and get proper SR component id for current install's referral code.

Reply:
Changed to use component for getting mapping table.